src/function/scalar/date/time_bucket.cpp

```diff
@@ -0,0 +1,387 @@
+		}
+		ts = Interval::Add(ts, Interval::Invert(offset));
+		int32_t ts_months = (Date::ExtractYear(ts) - 1970) * 12 + Date::ExtractMonth(ts) - 1;
+		// set origin 2000-01-01 00:00:00 for TimescaleDB compatibility
+		// there are 360 months between 1970-01-01 00:00:00 and 2000-01-01 00:00:00
+		int32_t origin_months = 360;
+
+		int32_t result_months = TimeBucketCommonWidthMoreThanMonths(bucket_width.months, ts_months, origin_months);
+		int32_t year =
+		    (result_months < 0 && result_months % 12 != 0) ? 1970 + result_months / 12 - 1 : 1970 + result_months / 12;
+		int32_t month =
+		    (result_months < 0 && result_months % 12 != 0) ? result_months % 12 + 13 : result_months % 12 + 1;
+		return Interval::Add(Cast::template Operation<date_t, TR>(Date::FromDate(year, month, 1)), offset);
+	}
+};
+
+struct TimeBucketOffsetTernaryOperator {
+	template <class TA, class TB, class TC, class TR>
+	static inline TR Operation(TA bucket_width, TB ts, TC offset) {
+		if (bucket_width.months == 0) {
+			int64_t bucket_width_micros = Interval::GetMicro(bucket_width);
+			if (bucket_width_micros <= 0) {
+				throw NotImplementedException("Period must be greater than 0");
+			}
+			return TimeBucketOffsetWidthLessThanDaysTernaryOperator::Operation<TA, TB, TC, TR>(bucket_width, ts,
+			                                                                                   offset);
+		} else if (bucket_width.months != 0 && bucket_width.days == 0 && bucket_width.micros == 0) {
+			if (bucket_width.months < 0) {
+				throw NotImplementedException("Period must be greater than 0");
+			}
+			return TimeBucketOffsetWidthMoreThanMonthsTernaryOperator::Operation<TA, TB, TC, TR>(bucket_width, ts,
+			                                                                                     offset);
+		} else {
+			throw NotImplementedException("Month intervals cannot have day or time component");
+		}
+	}
+};
+
+struct TimeBucketOriginWidthLessThanDaysTernaryOperator {
+	template <class TA, class TB, class TC, class TR>
+	static inline TR Operation(TA bucket_width, TB ts, TC origin) {
+		if (!Value::IsFinite(ts)) {
+			return Cast::template Operation<TB, TR>(ts);
+		}
+		int64_t bucket_width_micros = Interval::GetMicro(bucket_width);
+		int64_t ts_micros = Timestamp::GetEpochMicroSeconds(Cast::template Operation<TB, timestamp_t>(ts));
+		int64_t origin_micros = Timestamp::GetEpochMicroSeconds(Cast::template Operation<TB, timestamp_t>(origin));
+
+		return Cast::template Operation<timestamp_t, TR>(Timestamp::FromEpochMicroSeconds(
+		    TimeBucketCommonWidthLessThanDays(bucket_width_micros, ts_micros, origin_micros)));
+	}
+};
+
+struct TimeBucketOriginWidthMoreThanMonthsTernaryOperator {
+	template <class TA, class TB, class TC, class TR>
+	static inline TR Operation(TA bucket_width, TB ts, TC origin) {
+		if (!Value::IsFinite(ts)) {
+			return Cast::template Operation<TB, TR>(ts);
+		}
+		int32_t ts_months = (Date::ExtractYear(ts) - 1970) * 12 + Date::ExtractMonth(ts) - 1;
```

Comment:
There seems to be a common renormalisation here that is used in multiple places. Maybe it should be packaged up?